### PR TITLE
mask and experts list

### PR DESCRIPTION
**Pull request by Co1lin**
- mask some tensors of tokens for fmoe forward
- pass a list of expert classes to specify what experts in what order want to use

### CODE REVIEW

File: fmoe/layers.py [hunk 1]
```diff
@@ -196,14 +204,33 @@ def forward(self, inp):
 
         gate_top_k_idx, gate_score = self.gate(inp)
 
-        x = _fmoe_general_global_forward(
-            inp, 
+        # delete masked tensors
+        if self.mask != None and self.mask_dict != None:
```
Comment (laekov) [inline fmoe/layers.py:208]: plz use `is not None` instead of `!= None`

File: fmoe/layers.py [hunk 2]
```diff
@@ -196,14 +204,33 @@ def forward(self, inp):
...
         )
 
-        x = x.view(inp.shape[0], self.top_k, self.d_model)
-        gate_score = gate_score.view(inp.shape[0], 1, self.top_k)
+        # recover deleted tensors
+        if self.mask != None and self.mask_dict != None:
+            # to: (BxL') x top_k x d_model
+            fwd = fwd.view(-1, self.top_k, self.d_model)
+            # to: (BxL) x top_k x d_model
+            x = torch.zeros(mask.shape[0], self.top_k, self.d_model)
```
Comment (laekov) [inline fmoe/layers.py:225]: device=?, dtype=?

**Review by laekov — CHANGES_REQUESTED** [review]
Looks good, but a few details need to be updated or double checked.

File: fmoe/layers.py (hunk 1 above)
Comment (Co1lin) [reply fmoe/layers.py]: fixed in the new commit

File: fmoe/layers.py (hunk 2 above)
Comment (Co1lin) [reply fmoe/layers.py]: fixed in the new commit

**Review by laekov — APPROVED** [review]
LGTM. Thanks for contribution.
